ros/src/common/libs/state_machine_lib/src/state_context.cpp

```diff
@@ -23,358 +23,334 @@
+bool StateContext::setCallback(const CallbackType& _type, const std::string& _state_name,
+                               const std::function<void(const std::string&)>& _f)
 {
-  std::vector<BaseState *> ret;
-  for (uint64_t mask = STATE_SUB_END; _state_num_set != 0 && mask != 0; mask >>= 1)
+  bool ret = false;
+  uint64_t _state_id = getStateIDbyName(_state_name);
+  if (getStatePtr(_state_id))
   {
-    if (mask & _state_num_set)
+    switch (_type)
     {
-      ret.push_back(getStateObject(mask));
-      _state_num_set &= ~mask;
+      case CallbackType::UPDATE:
+        getStatePtr(_state_id)->setCallbackUpdate(_f);
+        ret = true;
+        break;
+      case CallbackType::ENTRY:
+        getStatePtr(_state_id)->setCallbackEntry(_f);
+        ret = true;
+        break;
+      case CallbackType::EXIT:
+        getStatePtr(_state_id)->setCallbackExit(_f);
+        ret = true;
+        break;
+      default:
+        break;
     }
   }
   return ret;
 }
+/*****************************/
+/*  Callback                 */
+/*****************************/
 
-void StateContext::update(void)
+void StateContext::onUpdate(void)
 {
-  for (auto &p : HolderMap)
-  {
-    if (p.first == BEHAVIOR_STATE)
-    {
-      for (auto &&state : getMultipleStates(p.second))
-      {
-        state->update();
-      }
-    }
-    else
-    {
-      if (p.second)
-        if (getStateObject(p.second))
-          getStateObject(p.second)->update();
-    }
-  }
+  root_state_->onUpdate();
 }
 
-void StateContext::inState(uint8_t _kind, uint64_t _prev_state_num)
+bool StateContext::isCurrentState(const std::string& state_name)
 {
-  if (_kind > UNKNOWN_STATE)
-  {
-    return;
-  }
-  else if (_kind == BEHAVIOR_STATE)
+  bool ret = false;
+  std::shared_ptr<State> state = root_state_;
+  do
   {
-    for (auto &&state : getMultipleStates(HolderMap[_kind]))
+    if (state->getStateName() == state_name)
     {
-      if (!(_prev_state_num & getStateNum(state)))
-        state->inState();
+      ret = true;
     }
-  }
-  else
-  {
-    if (getStateObject(HolderMap[_kind]))
-      getStateObject(HolderMap[_kind])->inState();
-  }
+    state = state->getChild();
+  } while (state != nullptr);
+  return ret;
 }
 
-BaseState *StateContext::getStateObject(const uint64_t &_state_num)
+void StateContext::nextState(const std::string& transition_key)
 {
-  if (_state_num)
+  const std::string dot_output_name = "/tmp/a.dot";
+  std::shared_ptr<State> state = root_state_;
+  std::string _target_state_name;
+  std::vector<std::string> key_list_;
+
+  while (state)
   {
-    if (StateStores[_state_num])
+    if (state->getTransitionMap().count(transition_key) != 0)
     {
-      return StateStores[_state_num];
-    }
-  }
-  return nullptr;
-}
+      const uint64_t transition_state_id = state->getTransitionMap().at(transition_key);
+      _target_state_name = state_map_[transition_state_id]->getStateName();
```

Comment:
For the same reason as above, you should use "at()" for element access.
```suggestion
      _target_state_name = state_map_.at(transition_state_id)->getStateName();
```

Reply:
Fixed.